### ISSUE TITLE
Define x/TokenFactory BeforeSend hook sentinel error

### ISSUE THREAD

**Issue by MirelEthernal** [osmosis-labs/osmosis#2570]
The BeforeSend hook implementation returns an error gained from the smart contract execution and depends on the info provided with the SC implementation.

**Artifacts:**

- osmosis/cosmos sdk fork at commit hash: [6fcd25f](https://github.com/osmosis-labs/cosmos-sdk/commit/6fcd25f8abb82ac4db337d72510429568362d348)
- Full Token Factory [branch](https://github.com/osmosis-labs/osmosis/tree/fullpowered-tokenfactory) holding new features

### Analysis summary:
https://github.com/osmosis-labs/osmosis/blob/f09305e60b5b23fec761ea14446ee33556313e69/x/tokenfactory/keeper/beforesend.go#L105-L110

**Suggestion:**
Token factory module should ensure that the error returned from the module holds a unique error code, that will explain what happened.
- Define a sentinel error  that will hold information about the denomination triggering the hook
- and wrap the error returned from the smart contract as well, in order to get the information about the reason - if provided in the smart contract (blocked from/to account address; the contract is frozen,...).

**Comment by MirelEthernal** [comment]
Tagging audit collaboration team, to review issues as agreed. @ValarDragon @sunnya97

**Comment by ValarDragon** [comment]
Agreed, we should wrap these error types, and make more meaningful wrapping info to indicate this is the source

Nice point!

**Comment by mattverse** [comment]
hey @alexanderbez are you stabbing this rn? If not, I'd like to PR this change with the other audit items!
